### ISSUE TITLE
Merging with NextGraph

### ISSUE THREAD

**Issue by joepio** [ontola/atomic-server#1012]
I had a couple of great conversations with @niko-ng who's working on [NextGraph](https://github.com/nextgraph-org). We've been exploring some ways in which we could work together towards something greater by collaborating. Niko proposed using NextGraph in AtomicServer.

This would be very impactful, but also worth exploring.

_This document will be updated / changed as I have more conversations with Niko_

## Why it would be cool (new features we gain)

- **Local-first**. That's a dream that I've had for a while. It's one of these characteristics that AtomicData doesn't have that would be very powerful, but is very hard to implement. NextGraph already has a lot of answers in this domain. #998 
- **Encrypted**. NextGraph encrypts everything. #300 
- **Branching**. NextGraph is a CRDT that allows multiple concurrent branches. This is almost there.
- **Drafts**. #129 
- **Real arrays**. With insert / splice methods etc. Nice. 
- **Form submissions**. In nextgraph, every document has an `inbox` where you can submit things. These things can be configured to have to conform to some shape. #875 

## NextGraph would replace...

- Atomic Commits as an abstraction layer (NextGraph supports multiple CRDTs). This also includes versioning, history, websockets, subscribing...
- Sled as KV store with RocksDB or some successor. We still need to investigate what this will be. But NextGraph will provide disk persistence. 
- Our existing Query index mechanisms. OxiGraph would do the heavy lifting, searching with SPARQL. However, that doesn't have lexicographic sorting or full text search.
- Atomic Agents. NextGraph also has a crypto based user / agent model. Wants to use [Opaque](https://blog.cloudflare.com/opaque-oblivious-passwords/), retrieves a wallet (containing your keys) safely from a server. You can open that wallet in a bunch of different ways. NextGraph's Pazzle abstraction will soon be replaced with a username-password. NextGraph would provide the login screen.
- Invites
- How URLs / subjects are built. Resources in NextGraph have special properties which link `did`s to (human readable) URL paths. 
- Schema validation (Atomic Schema). In NextGraph there is this thing called the "Verifier". It's a crate that runs verifications to make sure data is still valid. This should a new plugin, like an Atomic Schema validator plugin. We should consider doing JSON schema with NextGraph and extending that a bit. Will be a rust plugin.
- Authorization / persmissions. Atomic has read/write rights, as does NextGraph. 

## AtomicServer would still...

- Convert NextGraph resources to HTTP urls, similar to how ActivityPods are related to NextGraph right now
- Do pagination / indexing. The AtomicServer would be a Client which has its own agent, which has its own rights. The user can give access to _subsets_ of their data.
- Do full-text search using Tantivy
- Provide the front-end + libraries. 

## Questions & Doubts

- How does Atomic Schema relate to what NextGraph is doing?
- Content Addressing?
- P2P? NextGraph.
- The performance will almost certainly take a _big_ hit. There is no lexicographic sorting in Oxygraph at all, which means collections will be far slower. In addition to that, NextGraph is pretty slow when constructing a resource from commits. To what extent can we get good performance if we merge? What needs to change?
- NextGraph doesn't support querying over JSON yet, but it does support RDF indexing. Maybe we can index Atomic Data as RDF 
- NextGraph doesn't have fulltext search yet. Niko considered using Sonic, I told him about Tantivy, which I'd recommend because of some more powerful features. We could provide that with Atomic.
- Testing coverage of NextGraph is nowhere near that of AtomicServer. This doesn't mean the code is unreliable, but it does limit the extent to which I can mess around in it without feeling uncertain of it.
- Would it work with the Actix (tokio) runtime? NextGraph works with async-std, but that's no longer properly maintained. I'm not sure if one process can have both runtimes in parallel. [Relevant issue](https://git.nextgraph.org/NextGraph/nextgraph-rs/issues/41).
- The existing NextGraph wallet is kind of awkward to use, regarding sign-in / registration. This is planned to change, though.

## What would need to happen

- JS lib: use NG JS lib under the hood. That deals with fetching and editing data, keeping things in sync. The API of our SDKs should not change too much.
- Browser: change how users sign in. They'd use a NextGraph wallet. 
- Rust lib: should depend on the NG rust libraries. Ditch db.rs / query options / commit logic, replace with NG. Map nuri resources to HTTP(S) (or is this server terrain?). Change store initialization and persistance config to NG. Remove agents. 
- Server: make runtime compatible (tokio vs std_async). Ditch endpoints like /commit and /query. Keep /search.

**Comment by joepio** [comment]
We've received good news from NLNet: we'll receive a donation to work on local-first features :)

The first step will be to dive further into how AtomicServer and NextGraph will integrate.

**Comment by Tails** [comment]
Have you checked out TerminusDB? It's one of the few brancheable time-travelable rdf-graph-object based databases that would be able to support a few of these features outside of the app-layer.

**Comment by joepio** [comment]
> Have you checked out TerminusDB? It's one of the few brancheable time-travelable rdf-graph-object based databases that would be able to support a few of these features outside of the app-layer.

Yeah, cool project, @Tails! Haven't really considered it as a tool to power AtomicServer, though. It seems like a powerful tool, but also very heavy weight so not something I'd like to depend too much on. Also, it's very much RDF-first and has its own schema solution that doesn't seem compatible with Atomic.

**Comment by joepio** [comment]
We (@niko-ng and @Polleps) just discussed how we'll approach the NG + AS merge:

- Prioritize the JS side of things. We don't need server integration initially, we focus on front-end first.
- NG lives as a dependency of @tomic/lib. We should aim to keep the existing @tomic/react API. 
- Allow users to create NG resources. Maybe a boolean in the new resource page. This is temporary, just to explore NG + AS. There are now two stores: Atomic HTTP and NG. Depending on the schema (`http` vs `did:ng`) different logic is used to store, fetch resources. 
- We'll switch our text editor to NG's BlockNote

We'll write down most details in the [NG Merge Report](https://docs.google.com/document/d/1jTO6yi_r4-htT61YOvLL8difBq1iJiR_ghiv6ZfzIPE/edit?tab=t.0#heading=h.ys29et9pk330) (Google account access required)

**Comment by joepio** [comment]
We've opted for a bridge: https://github.com/ontola/atomic-ng-bridge
